File: crates/codegraph-core/src/extractors/fsharp.rs
use tree_sitter::{Node, Tree};
use crate::cfg::build_function_cfg;
use crate::complexity::compute_all_metrics;
use crate::types::*;
use super::helpers::*;
use super::SymbolExtractor;

pub struct FSharpExtractor;

impl SymbolExtractor for FSharpExtractor {
    fn extract(&self, tree: &Tree, source: &[u8], file_path: &str) -> FileSymbols {
        let mut symbols = FileSymbols::new(file_path.to_string());
        walk_tree(&tree.root_node(), source, &mut symbols, match_fsharp_node);
        walk_ast_nodes_with_config(&tree.root_node(), source, &mut symbols.ast_nodes, &FSHARP_AST_CONFIG);
        symbols
    }
}

fn match_fsharp_node(node: &Node, source: &[u8], symbols: &mut FileSymbols, _depth: usize) {
    match node.kind() {
        "named_module" => handle_named_module(node, source, symbols),
        "module_defn" => handle_module_defn(node, source, symbols),
        "function_declaration_left" => handle_function_decl(node, source, symbols),
        "type_definition" => handle_type_def(node, source, symbols),
        "import_decl" => handle_import_decl(node, source, symbols),
        "application_expression" => handle_application(node, source, symbols),
        "dot_expression" => handle_dot_expression(node, source, symbols),
        "value_definition" => handle_value_definition(node, source, symbols),
        _ => {}
    }
}

/// Find the enclosing module name, walking up through any number of
/// `module_defn` (nested signature modules) wrappers before reaching the
/// top-level `named_module`. Returns the dotted path, e.g. `Outer.Inner`.
///
/// Source files use `named_module` for the top-level `module M = …` and
/// the signature grammar (cargo 0.3.0) wraps nested signature modules in
/// `module_defn` nodes. The WASM signature grammar currently emits ERROR
/// nodes for nested signature modules so we cannot recover qualification
/// there — tracked under #1161.
fn enclosing_module_name(node: &Node, source: &[u8]) -> Option<String> {
    let mut parts: Vec<String> = Vec::new();
    let mut current = node.parent();
    while let Some(p) = current {
        match p.kind() {
            "module_defn" => {
                if let Some(id) = find_child(&p, "identifier") {
                    parts.push(node_text(&id, source).to_string());
                }
            }
            "named_module" => {
                if let Some(id) = find_child(&p, "long_identifier") {
                    parts.push(node_text(&id, source).to_string());
                }
                break;
            }
            _ => {}
        }
        current = p.parent();
    }
    if parts.is_empty() {
        return None;
    }
    parts.reverse();
    Some(parts.join("."))
}

fn handle_named_module(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let name_node = match find_child(node, "long_identifier") {
        Some(n) => n,
        None => return,
    };
    symbols.definitions.push(Definition {
        name: node_text(&name_node, source).to_string(),
        kind: "module".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
}

/// Handle nested signature modules (`module Foo = ...`) emitted by the
/// cargo 0.3.0 grammar as `module_defn`. Emits a `module` definition with
/// the dotted parent path (e.g. `Outer.Foo`) and lets the DFS walker
/// continue into child `val` declarations, which pick up the same path via
/// `enclosing_module_name`.
fn handle_module_defn(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let name_node = match find_child(node, "identifier") {
        Some(n) => n,
        None => return,
    };
    let raw = node_text(&name_node, source).to_string();
    // `enclosing_module_name` walks `node.parent()` upward, so calling it on
    // the `module_defn` itself yields the dotted prefix of its enclosing
    // module(s) without including this module's own name.
    let qualified = match enclosing_module_name(node, source) {
        Some(prefix) if !prefix.is_empty() => format!("{}.{}", prefix, raw),
        _ => raw,
    };
    symbols.definitions.push(Definition {
        name: qualified,
        kind: "module".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
}

fn handle_function_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    // function_declaration_left: first child is the function name identifier,
    // followed by argument_patterns.
    let name_node = match find_child(node, "identifier") {
        Some(n) => n,
        None => return,
    };
    let raw_name = node_text(&name_node, source).to_string();
    let line = start_line(node);

    // Avoid duplicates — the DFS walk also visits the inner curried
    // `function_declaration_left` of multi-parameter functions
    // (e.g. `let add x y = …`), which would otherwise push the same
    // `(name, line)` definition twice. Mirrors the JS extractor's guard,
    // which compares against the raw (unqualified) identifier text.
    if symbols
        .definitions
        .iter()
        .any(|d| d.name == raw_name && d.line == line)
    {
        return;
    }

    let module_name = enclosing_module_name(node, source);
    let qualified = match module_name {
        Some(m) => format!("{}.{}", m, raw_name),
        None => raw_name,
    };

    let params = extract_fsharp_params(node, source);

    // JS extractor uses the parent's endLine (the function_or_value_defn) for
    // a tighter bound; do the same to preserve parity.
    let end = node.parent().unwrap_or(*node);

    symbols.definitions.push(Definition {
        name: qualified,
        kind: "function".to_string(),
        line,
        end_line: Some(end_line(&end)),
        decorators: None,
        complexity: compute_all_metrics(&end, source, "fsharp"),
        cfg: build_function_cfg(&end, "fsharp", source),
        children: opt_children(params),
    });
}

fn extract_fsharp_params(decl_left: &Node, source: &[u8]) -> Vec<Definition> {
    let mut params = Vec::new();
    if let Some(arg_patterns) = find_child(decl_left, "argument_patterns") {
        collect_param_identifiers(&arg_patterns, source, &mut params);
    }
    params
}

fn collect_param_identifiers(node: &Node, source: &[u8], params: &mut Vec<Definition>) {
    if node.kind() == "identifier" {
        params.push(child_def(
            node_text(node, source).to_string(),
            "parameter",
            start_line(node),
        ));
        return;
    }
    for i in 0..node.child_count() {
        if let Some(child) = node.child(i) {
            collect_param_identifiers(&child, source, params);
        }
    }
}

fn handle_type_def(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    // type_definition contains union_type_defn, record_type_defn, etc.
    for i in 0..node.child_count() {
        let child = match node.child(i) {
            Some(c) => c,
            None => continue,
        };
        let kind = child.kind();
        if !matches!(
            kind,
            "union_type_defn"
                | "record_type_defn"
                | "type_abbreviation_defn"
                | "class_type_defn"
                | "interface_type_defn"
                | "type_defn"
        ) {
            continue;
        }

        let name = match find_child(&child, "type_name") {
            Some(type_name) => find_child(&type_name, "identifier")
                .map(|n| node_text(&n, source).to_string())
                .unwrap_or_else(|| node_text(&type_name, source).to_string()),
            None => match find_child(&child, "identifier") {
                Some(id) => node_text(&id, source).to_string(),
                None => continue,
            },
        };

        let mut children: Vec<Definition> = Vec::new();
        extract_type_members(&child, source, &mut children);

        symbols.definitions.push(Definition {
            name,
            kind: determine_type_kind(kind).to_string(),
            line: start_line(&child),
            end_line: Some(end_line(&child)),
            decorators: None,
            complexity: None,
            cfg: None,
            children: opt_children(children),
        });
    }
}

fn determine_type_kind(node_kind: &str) -> &'static str {
    match node_kind {
        "union_type_defn" => "enum",
        "record_type_defn" => "record",
        "class_type_defn" => "class",
        "interface_type_defn" => "interface",
        _ => "type",
    }
}

fn extract_type_members(type_defn: &Node, source: &[u8], children: &mut Vec<Definition>) {
    for i in 0..type_defn.child_count() {
        let child = match type_defn.child(i) {
            Some(c) => c,
            None => continue,
        };

        match child.kind() {
            "union_type_case" => {
                if let Some(name) = find_child(&child, "identifier") {
                    children.push(child_def(
                        node_text(&name, source).to_string(),
                        "property",
                        start_line(&child),
                    ));
                }
            }
            "record_field" => {
                let name_node = child
                    .child_by_field_name("name")
                    .or_else(|| find_child(&child, "identifier"));
                if let Some(name) = name_node {
                    children.push(child_def(
                        node_text(&name, source).to_string(),
                        "property",
                        start_line(&child),
                    ));
                }
            }
            // Recurse into container nodes that hold cases/fields.
            "union_type_cases" | "record_fields" => {
                extract_type_members(&child, source, children);
            }
            _ => {}
        }
    }
}

fn handle_import_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let module_node = match find_child(node, "long_identifier") {
        Some(n) => n,
        None => return,
    };

    let source_name = node_text(&module_node, source).to_string();
    let last = source_name
        .split('.')
        .last()
        .unwrap_or(&source_name)
        .to_string();

    symbols
        .imports
        .push(Import::new(source_name, vec![last], start_line(node)));
}

fn handle_application(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let func_node = match node.child(0) {
        Some(n) => n,
        None => return,
    };

    // Mirrors the JS extractor's `handleApplication`: the full dotted name
    // (e.g. `Service.createUser`) is stored in `name`. Splitting `name` into
    // `(receiver, method)` would diverge from the JS engine's output and
    // change which resolution rules fire downstream.
    match func_node.kind() {
        "identifier" | "long_identifier" => {
            symbols.calls.push(Call {
                name: node_text(&func_node, source).to_string(),
                line: start_line(node),
                dynamic: None,
                receiver: None,
            });
        }
        "long_identifier_or_op" => {
            // Inner child is either `identifier` (bare, e.g. `validateUser`) or
            // `long_identifier` (qualified, e.g. `Repository.save`). Order
            // matches the JS extractor (`identifier` first). Operator forms
            // like `( + )` have neither child; we emit nothing in that case,
            // mirroring the JS extractor's silent skip.
            if let Some(inner) =
                find_first_child_of_types(&func_node, &["identifier", "long_identifier"])
            {
                symbols.calls.push(Call {
                    name: node_text(&inner, source).to_string(),
                    line: start_line(node),
                    dynamic: None,
                    receiver: None,
                });
            }
        }
        _ => {}
    }
}

fn handle_dot_expression(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    // Mirrors the JS extractor's `handleDotExpression`: collect identifier
    // segments and emit `name = last`, `receiver = everything-before`.
    let mut parts: Vec<String> = Vec::new();
    for i in 0..node.child_count() {
        if let Some(child) = node.child(i) {
            match child.kind() {
                "identifier" | "long_identifier" => {
                    parts.push(node_text(&child, source).to_string());
                }
                _ => {}
            }
        }
    }
    if parts.len() >= 2 {
        let method = parts.last().cloned().unwrap_or_default();
        let receiver = parts[..parts.len() - 1].join(".");
        symbols.calls.push(Call {
            name: method,
            line: start_line(node),
            dynamic: None,
            receiver: Some(receiver),
        });
    }
}

/// Handle `val name : type` declarations in `.fsi` signature files.
///
/// The signature grammar reuses the `value_definition` node kind for `val`
/// declarations, distinguished from the source grammar's `let` bindings by
/// the first child being the literal `val` keyword. Source-file
/// `value_definition` nodes (which start with `let`) are intentionally
/// ignored here to preserve `.fs` extractor parity.
fn handle_value_definition(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let first = match node.child(0) {
        Some(c) => c,
        None => return,
    };
    if first.kind() != "val" {
        return;
    }

    let decl_left = match find_child(node, "value_declaration_left") {
        Some(n) => n,
        None => return,
    };
    let name = match extract_value_name(&decl_left, source) {
        Some(n) => n,
        None => return,
    };

    let kind = if has_function_type(node) { "function" } else { "variable" };
    let module_name = enclosing_module_name(node, source);
    let qualified = match module_name {
        Some(m) => format!("{}.{}", m, name),
        None => name,
    };

    symbols.definitions.push(Definition {
        name: qualified,
        kind: kind.to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
}

fn extract_value_name(decl_left: &Node, source: &[u8]) -> Option<String> {
    let pattern = find_child(decl_left, "identifier_pattern")?;
    let ident = find_child(&pattern, "long_identifier_or_op")
        .and_then(|n| find_child(&n, "identifier"))
        .or_else(|| find_child(&pattern, "identifier"))?;
    Some(node_text(&ident, source).to_string())
}

fn has_function_type(node: &Node) -> bool {
    // The grammar wraps every type signature in `curried_spec`. A function type
    // (e.g. `val add : int -> int -> int`) contains one or more `arguments_spec`
    // children; a plain value (e.g. `val pi : float`) wraps a single `simple_type`.
    let Some(curried) = find_child(node, "curried_spec") else { return false };
    for i in 0..curried.child_count() {
        if let Some(child) = curried.child(i) {
            if child.kind() == "arguments_spec" {
                return true;
            }
        }
    }
    false
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::extractors::SymbolExtractor;
    use tree_sitter::Parser;

    fn parse_source(code: &str) -> FileSymbols {
        let mut parser = Parser::new();
        parser
            .set_language(&tree_sitter_fsharp::LANGUAGE_FSHARP.into())
            .unwrap();
        let tree = parser.parse(code.as_bytes(), None).unwrap();
        FSharpExtractor.extract(&tree, code.as_bytes(), "test.fs")
    }

    fn parse_signature(code: &str) -> FileSymbols {
        let mut parser = Parser::new();
        parser
            .set_language(&tree_sitter_fsharp::LANGUAGE_SIGNATURE.into())
            .unwrap();
        let tree = parser.parse(code.as_bytes(), None).unwrap();
        FSharpExtractor.extract(&tree, code.as_bytes(), "test.fsi")
    }

    #[test]
    fn signature_extracts_val_declarations() {
        let s = parse_signature("namespace MyApp.Domain\n\nval add : int -> int -> int\nval pi : float\n");
        let add = s
            .definitions
            .iter()
            .find(|d| d.name == "add")
            .expect("val add should be extracted");
        assert_eq!(add.kind, "function");
        let pi = s
            .definitions
            .iter()
            .find(|d| d.name == "pi")
            .expect("val pi should be extracted");
        assert_eq!(pi.kind, "variable");
    }

    #[test]
    fn signature_extracts_bare_val_declarations() {
        let s = parse_signature("val negate : int -> int\nval count : int\n");
        assert!(s
            .definitions
            .iter()
            .any(|d| d.name == "negate" && d.kind == "function"));
        assert!(s
            .definitions
            .iter()
            .any(|d| d.name == "count" && d.kind == "variable"));
    }

    #[test]
    fn source_grammar_does_not_extract_let_bindings_as_val() {
        // `let x = 5` is a value_definition in the source grammar but its
        // first child is `let`, not `val`. Our handler must not extract it
        // (preserves prior `.fs` extraction parity — only function_declaration_left
        // produces definitions in source files).
        let s = parse_source("module M\n\nlet x = 5\n");
        assert!(
            s.definitions.iter().all(|d| d.name != "x"),
            "let bindings in .fs files must not be extracted as val definitions"
        );
    }

    #[test]
    fn signature_qualifies_val_inside_nested_module_defn() {
        // The cargo 0.3.0 signature grammar wraps `module Foo = ...` as a
        // `module_defn` node (the WASM 0.1.0 grammar emits ERROR for this
        // construct — tracked under #1161). The `val` declarations inside
        // must be qualified with the module path.
        let s = parse_signature("namespace X\n\nmodule Foo =\n  val add : int -> int\n");
        assert!(
            s.definitions.iter().any(|d| d.name == "Foo.add" && d.kind == "function"),
            "val add nested under `module Foo =` must be indexed as `Foo.add`, got: {:?}",
            s.definitions.iter().map(|d| &d.name).collect::<Vec<_>>(),
        );
        assert!(
            s.definitions.iter().any(|d| d.name == "Foo" && d.kind == "module"),
            "module Foo must be indexed as a module definition"
        );
    }

    #[test]
    fn source_grammar_does_not_extract_val_mutable_class_fields() {
        // `val mutable count: int = 0` inside a class is parsed as a `member_defn`
        // node in the source grammar — NOT a `value_definition` — so our
        // `value_definition`/`val`-first-child handler does not see it.
        // This regression guard makes that empirical fact explicit.
        let s = parse_source(
            "module M\n\ntype C() =\n    val mutable count: int = 0\n",
        );
        assert!(
            s.definitions.iter().all(|d| d.name != "count"),
            "val mutable class fields must not be extracted by the signature value_definition handler"
        );
    }
}

File: src/extractors/fsharp.ts
import type {
  Call,
  ExtractorOutput,
  SubDeclaration,
  TreeSitterNode,
  TreeSitterTree,
} from '../types.js';
import { findChild, nodeEndLine } from './helpers.js';

/**
 * Extract symbols from F# files.
 *
 * Grammar source: `tree-sitter-fsharp` v0.3.0 installed via a pinned GitHub
 * tarball in `package.json` because the ionide/tree-sitter-fsharp project has
 * no v0.3.0 release published to the npm registry. The cargo crate the native
 * engine uses is also v0.3.0; both engines must stay aligned. Upgrading
 * requires a manual edit of the tarball URL in `package.json` and
 * `package-lock.json` — `npm update` will not bump this entry.
 *
 * tree-sitter-fsharp grammar notes:
 * - named_module: top-level module declaration
 * - function_declaration_left: LHS of `let name params = ...`
 * - import_decl: `open Namespace`
 * - type_definition > union_type_defn / record_type_defn
 * - application_expression: function calls
 */
export function extractFSharpSymbols(tree: TreeSitterTree, _filePath: string): ExtractorOutput {
  const ctx: ExtractorOutput = {
    definitions: [],
    calls: [],
    imports: [],
    classes: [],
    exports: [],
    typeMap: new Map(),
  };

  walkFSharpNode(tree.rootNode, ctx, null);
  return ctx;
}

function walkFSharpNode(
  node: TreeSitterNode,
  ctx: ExtractorOutput,
  currentModule: string | null,
): void {
  let nextModule = currentModule;

  switch (node.type) {
    case 'named_module':
      nextModule = handleNamedModule(node, ctx);
      break;
    case 'module_defn':
      // Nested signature module (`module Foo = ...`) in `.fsi` files,
      // emitted by both the WASM (npm ionide tarball v0.3.0) and cargo
      // v0.3.0 tree-sitter-fsharp signature grammars. Accumulate the
      // dotted module path so nested `val` declarations are qualified
      // as `Outer.Inner.foo` in parity with the native engine.
      nextModule = handleModuleDefn(node, ctx, currentModule);
      break;
    case 'function_declaration_left':
      handleFunctionDecl(node, ctx, currentModule);
      break;
    case 'type_definition':
      handleTypeDef(node, ctx);
      break;
    case 'import_decl':
      handleImportDecl(node, ctx);
      break;
    case 'application_expression':
      handleApplication(node, ctx);
      break;
    case 'dot_expression':
      handleDotExpression(node, ctx);
      break;
    case 'value_definition':
      handleValueDefinition(node, ctx, currentModule);
      break;
  }

  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child) walkFSharpNode(child, ctx, nextModule);
  }
}

function handleNamedModule(node: TreeSitterNode, ctx: ExtractorOutput): string | null {
  const nameNode = findChild(node, 'long_identifier');
  if (!nameNode) return null;

  ctx.definitions.push({
    name: nameNode.text,
    kind: 'module',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });

  return nameNode.text;
}

function handleModuleDefn(
  node: TreeSitterNode,
  ctx: ExtractorOutput,
  currentModule: string | null,
): string | null {
  // `module_defn` (cargo 0.3.0 signature grammar) wraps `module Foo = ...`
  // sections inside an outer `namespace` or another module. The name is a
  // direct `identifier` child.
  const nameNode = findChild(node, 'identifier');
  if (!nameNode) return currentModule;

  const qualified = currentModule ? `${currentModule}.${nameNode.text}` : nameNode.text;
  ctx.definitions.push({
    name: qualified,
    kind: 'module',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
  return qualified;
}

function handleFunctionDecl(
  node: TreeSitterNode,
  ctx: ExtractorOutput,
  currentModule: string | null,
): void {
  // function_declaration_left: "add x y" — first child is the name identifier
  const nameNode = findChild(node, 'identifier');
  if (!nameNode) return;

  // Avoid duplicates — the walk will also visit children
  if (
    ctx.definitions.some((d) => d.name === nameNode.text && d.line === node.startPosition.row + 1)
  )
    return;

  const params = extractFSharpParams(node);
  const name = currentModule ? `${currentModule}.${nameNode.text}` : nameNode.text;

  ctx.definitions.push({
    name,
    kind: 'function',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node.parent ?? node),
    children: params.length > 0 ? params : undefined,
  });
}

function extractFSharpParams(declLeft: TreeSitterNode): SubDeclaration[] {
  const params: SubDeclaration[] = [];
  const argPatterns = findChild(declLeft, 'argument_patterns');
  if (!argPatterns) return params;

  collectParamIdentifiers(argPatterns, params);
  return params;
}

function collectParamIdentifiers(node: TreeSitterNode, params: SubDeclaration[]): void {
  if (node.type === 'identifier') {
    params.push({ name: node.text, kind: 'parameter', line: node.startPosition.row + 1 });
    return;
  }
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child) collectParamIdentifiers(child, params);
  }
}

function handleTypeDef(node: TreeSitterNode, ctx: ExtractorOutput): void {
  // type_definition contains union_type_defn, record_type_defn, etc.
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (!child) continue;

    if (
      child.type === 'union_type_defn' ||
      child.type === 'record_type_defn' ||
      child.type === 'type_abbreviation_defn' ||
      child.type === 'class_type_defn' ||
      child.type === 'interface_type_defn' ||
      child.type === 'type_defn'
    ) {
      const nameNode = findChild(child, 'type_name');
      const name = nameNode
        ? (findChild(nameNode, 'identifier')?.text ?? nameNode.text)
        : findChild(child, 'identifier')?.text;
      if (!name) continue;

      const kind = determineFSharpTypeKind(child);
      const children: SubDeclaration[] = [];
      extractFSharpTypeMembers(child, children);

      ctx.definitions.push({
        name,
        kind,
        line: child.startPosition.row + 1,
        endLine: nodeEndLine(child),
        children: children.length > 0 ? children : undefined,
      });
    }
  }
}

function determineFSharpTypeKind(
  typeDefn: TreeSitterNode,
): 'class' | 'type' | 'record' | 'enum' | 'interface' {
  switch (typeDefn.type) {
    case 'union_type_defn':
      return 'enum';
    case 'record_type_defn':
      return 'record';
    case 'class_type_defn':
      return 'class';
    case 'interface_type_defn':
      return 'interface';
    default:
      return 'type';
  }
}

function extractFSharpTypeMembers(typeDefn: TreeSitterNode, children: SubDeclaration[]): void {
  for (let i = 0; i < typeDefn.childCount; i++) {
    const child = typeDefn.child(i);
    if (!child) continue;

    if (child.type === 'union_type_case') {
      const nameNode = findChild(child, 'identifier');
      if (nameNode) {
        children.push({
          name: nameNode.text,
          kind: 'property',
          line: child.startPosition.row + 1,
        });
      }
    }
    if (child.type === 'record_field') {
      const nameNode = child.childForFieldName('name') || findChild(child, 'identifier');
      if (nameNode) {
        children.push({
          name: nameNode.text,
          kind: 'property',
          line: child.startPosition.row + 1,
        });
      }
    }
    // Recurse into containers like union_type_cases
    if (child.type === 'union_type_cases' || child.type === 'record_fields') {
      extractFSharpTypeMembers(child, children);
    }
  }
}

function handleImportDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const moduleNode = findChild(node, 'long_identifier');
  if (!moduleNode) return;

  const source = moduleNode.text;
  ctx.imports.push({
    source,
    names: [source.split('.').pop() || source],
    line: node.startPosition.row + 1,
  });
}

function handleApplication(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const funcNode = node.child(0);
  if (!funcNode) return;

  if (funcNode.type === 'identifier' || funcNode.type === 'long_identifier') {
    ctx.calls.push({ name: funcNode.text, line: node.startPosition.row + 1 });
  } else if (funcNode.type === 'long_identifier_or_op') {
    const id = findChild(funcNode, 'identifier') || findChild(funcNode, 'long_identifier');
    if (id) ctx.calls.push({ name: id.text, line: node.startPosition.row + 1 });
  }
}

function handleDotExpression(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const parts: string[] = [];
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child && (child.type === 'identifier' || child.type === 'long_identifier')) {
      parts.push(child.text);
    }
  }
  if (parts.length >= 2) {
    const call: Call = {
      name: parts[parts.length - 1]!,
      receiver: parts.slice(0, -1).join('.'),
      line: node.startPosition.row + 1,
    };
    ctx.calls.push(call);
  }
}

// Handle `val name : type` declarations in `.fsi` signature files.
// The signature grammar reuses `value_definition` for `val` bindings,
// distinguished from the source grammar's `let` bindings by the first
// child being the literal `val` keyword. Source-file `value_definition`
// nodes (which start with `let`) are intentionally ignored to preserve
// `.fs` extractor parity.
function handleValueDefinition(
  node: TreeSitterNode,
  ctx: ExtractorOutput,
  currentModule: string | null,
): void {
  const first = node.child(0);
  if (!first || first.type !== 'val') return;

  const declLeft = findChild(node, 'value_declaration_left');
  if (!declLeft) return;

  const pattern = findChild(declLeft, 'identifier_pattern');
  if (!pattern) return;

  const ident =
    findChild(findChild(pattern, 'long_identifier_or_op') ?? pattern, 'identifier') ??
    findChild(pattern, 'identifier');
  if (!ident) return;

  // The npm and cargo tree-sitter-fsharp 0.3.0 grammars — though sharing a
  // version tag — emit type signatures with different node shapes:
  //   • WASM (npm 0.3.0 ionide tarball): `function_type` is the explicit
  //     function-type kind, present as a direct child of `value_definition`
  //     for `a -> b` types; plain values (e.g. `val pi : float`) appear as
  //     `simple_type`.
  //   • Native (cargo 0.3.0): every type signature is wrapped in
  //     `curried_spec`. A function type contains one or more `arguments_spec`
  //     children; a plain value wraps a single `simple_type`.
  // Classify as a function whenever `function_type` appears OR a
  // `curried_spec` contains an `arguments_spec` child, so both engines stay
  // in parity until the grammars converge.
  let hasFunctionType = false;
  for (let i = 0; i < node.childCount; i++) {
    const c = node.child(i);
    if (!c) continue;
    if (c.type === 'function_type') {
      hasFunctionType = true;
      break;
    }
    if (c.type === 'curried_spec') {
      for (let j = 0; j < c.childCount; j++) {
        if (c.child(j)?.type === 'arguments_spec') {
          hasFunctionType = true;
          break;
        }
      }
      if (hasFunctionType) break;
    }
  }

  const name = currentModule ? `${currentModule}.${ident.text}` : ident.text;
  ctx.definitions.push({
    name,
    kind: hasFunctionType ? 'function' : 'variable',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
}
